src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/HACK_NodeHelper.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using Microsoft.VisualStudio.Packaging;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands.Ordering
+{
+    /// <summary>
+    /// These are helper functions to select items, expand and collapse folders in a IVsUIHierarchy.
+    /// </summary>
+    internal static class HACK_NodeHelper
```

Comment:
We cannot check in this name. Can you expand what is causing this to be a hack?

Reply:
I'd like this to be pulled into a separate instance class/interface and for us to either:

1) Make these methods Async and switch to the UI thread at the start of them
2) Make these methods sync and enforce that we're on the UI thread - and every caller to deliberately switch to the UI thread.

Reply:
The reason for this - is it makes it very clear when/when we're not on the UI thread. At the moment, a simple refactoring/ConfigureAwait change could result in this being called on a background thread and introduce a COM RPC.

Reply:
I will add a comment describing this.

Basically, it's a hack because CPS should give us ways to do this. They have implementations for this functionality but do not expose it. But I need it in this case, and I have no choice but to implement it.

I can make them async and switch to the ui thread.

Reply:
Don't call it hack, just give it a real name - and check it in. File a bug against CPS to open this up, and if they fund/get around to it, we'll remove our version.